### ISSUE TITLE
Changes to MvcOption's settings (SuppressInputFormatterBuffering & AllowBindingUndefinedValueToEnumType) are not taking affect

### ISSUE THREAD

**Issue by kichalla** [aspnet/Mvc#6858]
I found this bug today while working on a different issue. The following piece of code doesn't cause the affect that you would expect them to:

```c#
services.AddMvc(o =>
{
    o.AllowBindingUndefinedValueToEnumType = true; // was added post 2.0 release
    o.SuppressInputFormatterBuffering = true; // is part of 2.0 release
});
```

This is because the following pieces of code get executed even before the above option manipulation occurs causing the affects not to be seen:

https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcCoreMvcOptionsSetup.cs#L49
https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Formatters.Json/Internal/MvcJsonMvcOptionsSetup.cs#L80

**Workaround:**
Sequence the `Configure<MvcOptions>` method before the `AddMvc` method

```c#
services.Configure<MvcOptions>(o =>
{
    o.AllowBindingUndefinedValueToEnumType = true;
    o.SuppressInputFormatterBuffering = true;
});

services.AddMvc();
```

**Fix**: 
Pass around MvcOptions instance rather than the actual value. Any mutation to MvcOptions can be observed by individual components

cc @rynowak , @dougbu , @Eilon

**Comment by dougbu** [comment]
> Pass around `MvcOptions` instance rather than the actual value.

To clarify: Do you mean everywhere? We have a lot of `public` constructors with parameters taken from within `MvcOptions`. Would likely end up with at least a few deprecated constructors lying around (ick).

On the other hand, we could change the (also numerous) constructors that dereference `options.Value` before they return without such breaking changes.

Of course, fixing `AllowBindingUndefinedValueToEnumType` is easy since we haven't released the feature yet.

**Comment by kichalla** [comment]
> Do you mean everywhere?

May be not everywhere. For example, if the options instance is DI injected then we needn't worry as the same singleton instance is passed around (right?)

> We have a lot of public constructors with parameters taken from within MvcOptions. Would likely end up with at least a few deprecated constructors lying around (ick).

If the parameters are reference types then there is no problem. So choosing to either pass around MvcOptions or only the individual properties of it is error prone (with checking if they are reference type or not) and would make the code base inconsistent for us and the users

> On the other hand, we could change the (also numerous) constructors that dereference options.Value before they return without such breaking changes.

I am not clear about what you mean here

**Comment by dougbu** [comment]
> we needn't worry as the same singleton instance is passed around (right?)

This relates to my "deference options" point. Passing `XyzOptions` around is fine up until something reads a `struct` from it or otherwise locks the information. That "something" may be the constructor of a service.

> I am not clear about what you mean here

Yeah, just de-referencing `options.Value` from an `IOptions<XyzOptions>` parameter isn't enough. I meant reaching within `options.Value` to get something.

Note that `struct`s aren't the only potential problem here. For example, a constructor may evaluate an `IEnumberable<Zyx>` property in `XyzOptions` and create a `List<Zyz>`. The class containing such a constructor will ignore further changes to the `XyzOptions`'s property.

**Comment by Eilon** [comment]
Let's try to fix both of these in 2.1.

**Comment by Eilon** [comment]
@kichalla done?
